R/chunk_req_perform.R

```diff
@@ -0,0 +1,204 @@
+#'   list of parsed responses.
+#' @export
+#'
+#' @examples
+#' req <- request("https://api.restful-api.dev/objects")
+#'
+#' ids <- 1:7
+#' chunk_size <- 3
+#'
+#' apply_chunk <- function(req, chunk) {
+#'   chunk <- rlang::set_names(chunk, "id")
+#'   req_url_query(req, !!!chunk)
+#' }
+#'
+#' parse_resp <- function(resp) {
+#'   parsed <- resp_body_json(resp)
+#'   data.frame(
+#'     id = sapply(parsed, function(x) x$id),
+#'     name = sapply(parsed, function(x) x$name)
+#'   )
+#' }
+#'
+#' # in most cases `chunk_req_perform()` should be sufficient
+#' \dontrun{
+#' responses <- chunk_req_perform(
+#'   req,
+#'   chunk_size = 10,
+#'   data = ids,
+#'   apply_chunk = apply_chunk,
+#'   parse_resp = parse_resp
+#' )
+#' }
+#'
+#' # in case you need more control of how the requests are performed you can create
+#' # the requests first with `req_chunk()` ...
+#' requests <- req_chunk(
+#'   req,
+#'   chunk_size = 3,
+#'   data = ids,
+#'   apply_chunk = apply_chunk
+#' )
+#'
+#' # a simple list of requests
+#' requests
+#'
+#' # ... which you can perform with `req_perform()`
+#' \dontrun{
+#' n <- length(requests)
+#' paths <- character(n)
+#' responses <- list()
+#' for (i in seq2(1, n)) {
+#'   paths[[i]] <- tempfile()
+#'   responses[[i]] <- req_perform(requests[[i]], path = paths[[i]])
+#' }
+#' }
+req_chunk <- function(req,
+                      data,
+                      chunk_size,
+                      apply_chunk) {
+  check_request(req)
+  chunks <- vec_chop_by_size(data, chunk_size)
+  check_function2(apply_chunk, args = c("req", "chunk"))
+
+  n <- length(chunks)
+  requests <- lapply(chunks, function(chunk) apply_chunk(req, chunk))
+  new_chunked_request(requests, chunks)
+}
+
+#' @export
+#' @rdname req_chunk
+#' @inheritParams paginate_req_perform
+#' @param parse_resp A function with one argument `resp` that parses the
+#'   response.
+#' @param progress Whether to show a progress bar. Use `TRUE` to turn on a basic
+#'   progress bar, use a string to give it a name, or see progress_bars for more
+#'   details.
+chunk_req_perform <- function(req,
+                              data,
+                              chunk_size,
+                              apply_chunk,
+                              parse_resp = NULL,
+                              progress = TRUE,
+                              error_call = current_env()) {
+  chunked_requests <- req_chunk(
+    req = req,
+    data = data,
+    chunk_size = chunk_size,
+    apply_chunk = apply_chunk
+  )
+  requests <- chunked_requests$requests
+  the$last_chunks <- chunked_requests$chunks
+
+  parse_resp <- parse_resp %||% function(resp) resp
+  check_function2(parse_resp, args = "resp", call = error_call)
+
+  n <- length(requests)
+  the$last_chunked_responses <- vector("list", n)
+
+  pb <- create_progress_bar(total = n, name = "Request chunks", progress)
+  show_progress <- !is.null(pb)
+  env <- current_env()
+
+  for (i in seq2(1, n)) {
```

Comment:
It feels like this should share more implementation with `multi_req_perform()`. Maybe we should expose the `host_con` argument from `curl::multi_set()`?

Reply:
I think this is orthogonal to that problem, since I'm thinking more of just using `multi_req_perform()` here since you have a list of requests and vector of paths.